lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -3481,6 +3481,34 @@ bool NonLocalAnchor::updateModule(StringAttr oldMod, StringAttr newMod) {
   return updateMade;
 }
 
+bool NonLocalAnchor::updateModuleAndInnerRef(
+    StringAttr oldMod, StringAttr newMod,
+    const llvm::DenseMap<StringAttr, StringAttr> &renameMap) {
+  auto fromRef = FlatSymbolRefAttr::get(oldMod);
+  SmallVector<Attribute> namepathNew;
+  bool updateMade = false;
+  for (auto element : namepath()) {
+    if (auto innerRef = element.dyn_cast<hw::InnerRefAttr>()) {
+      if (innerRef.getModule() == oldMod) {
+        // If the leaf module updated, then the symbol should also be renamed.
+        auto to = renameMap.find(innerRef.getName());
+        assert(to != renameMap.end() && "should have been renamed");
+        updateMade = true;
+        namepathNew.push_back(hw::InnerRefAttr::get(newMod, to->second));
+      } else
+        namepathNew.push_back(element);
+    } else if (element == fromRef) {
+      updateMade = true;
+      namepathNew.push_back(FlatSymbolRefAttr::get(newMod));
+    } else {
+      namepathNew.push_back(element);
+    }
+  }
+  if (updateMade)
+    namepathAttr(ArrayAttr::get(getContext(), namepathNew));
+  return updateMade;
+}
+
```

Comment:
I still don't know what the purpose of the renameMap is, even after reading the whole function.